extensions/aws/processors/PutS3Object.cpp

```diff
@@ -77,7 +78,31 @@ void PutS3Object::onSchedule(const std::shared_ptr<core::ProcessContext> &contex
     use_virtual_addressing_ = !*use_path_style_access;
   }
 
+  context->getProperty(MultipartThreshold.getName(), multipart_threshold_);
+  if (multipart_threshold_ > getMaxUploadSize() || multipart_threshold_ < getMinPartSize()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Multipart Threshold is not between the valid 5MB and 5GB range!");
+  }
+  logger_->log_debug("PutS3Object: Multipart Threshold %" PRIu64, multipart_threshold_);
+  context->getProperty(MultipartPartSize.getName(), multipart_size_);
+  if (multipart_size_ > getMaxUploadSize() || multipart_size_ < getMinPartSize()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Multipart Part Size is not between the valid 5MB and 5GB range!");
+  }
+  logger_->log_debug("PutS3Object: Multipart Size %" PRIu64, multipart_size_);
+
+
+  multipart_upload_ageoff_interval_ = minifi::utils::getRequiredPropertyOrThrow<core::TimePeriodValue>(*context, MultipartUploadAgeOffInterval.getName()).getMilliseconds();
+  logger_->log_debug("PutS3Object: Multipart Upload Ageoff Interval %" PRIu64 " ms", multipart_upload_ageoff_interval_.count());
+
+  multipart_upload_max_age_threshold_ = minifi::utils::getRequiredPropertyOrThrow<core::TimePeriodValue>(*context, MultipartUploadMaxAgeThreshold.getName()).getMilliseconds();
+  logger_->log_debug("PutS3Object: Multipart Upload Max Age Threshold %" PRIu64 " ms", multipart_upload_max_age_threshold_.count());
+
   fillUserMetadata(context);
+
+  std::string multipart_temp_dir;
+  context->getProperty(TemporaryDirectoryMultipartState.getName(), multipart_temp_dir);
+
+
+  s3_wrapper_.initailizeMultipartUploadStateStorage(multipart_temp_dir, getUUIDStr());
```

Comment:
Why do we need a temp dir? We have standard state management with multiple backends.

Reply:
First I implemented it with the state manager, but there was a problem with that. The state manager only commits the state at session commit and if the process is killed or the upload fails and the session is rolled back then the states between the part uploads are lost. Even when I tried to commit the state manager manually it caused an exception at the end of the trigger because the session commit could not commit (or rollback in case of a failure) an already commited session. It may be the same issue in NiFi because they also use a separate temporary directory for this state management.

Reply:
I don't really like circumventing the state manager, we should aim to make it work by extending it instead IMO. Not necessarily in this PR, but that works, too. And I want to avoid introducing a property that will need to be removed later, because removing a processor property may break old config.ymls that contain it. (Unless dynamic properties are enabled, but they don't seem to make sense here.)

I think should go with another compromise, my ideas:
1. Give up the functionality of continuing failed multipart uploads after a restart. It could still continue if the agent wasn't restarted, but restart would restart the upload.
2. Use a system-wide temporary directory, e.g. /tmp/minifi-multipart-state/processor_uuid
3. Configure the multipart state dir globally in minifi.properties, because that doesn't break configs when we drop the config property

I gravitate towards 1, but I'm interested in your viewpoint.

Reply:
I agree that it would be optimal to use the state manager. In its current form however as a single transaction is tied to a single onTrigger session it will probably take a larger effort to figure out how should commit and rollback work with multiple states in a single session. Because of this I wouldn't make it part of this PR.

IMO we shouldn't give up continuing failed multipart uploads after a restart as it could be a useful feature for the user if for any reason the system is restarted, the process is killed or maybe the MiNiFi service is restarted. I would also like to make it configurable to avoid any directory permission issues or just to give the user the option to decide where MiNiFi should be allowed to write on the system. Because of this I would go with option 3.

Reply:
What would you default to, when it's not configured? OS tempdir, or MINIFI_HOME subdir?

Reply:
Added a commit with this change: ea2548e618b539facccf226fbc417c6d46fe71ee

Reply:
Added a jira ticket for the state management change https://issues.apache.org/jira/browse/MINIFICPP-2140